File: hydra/garaga/hints/eisenstein.py
"""
 -----------------------------------------------------------------------------
 ε – Eisenstein integers in pure Python (fully commented)
 -----------------------------------------------------------------------------
 A **beginner‑friendly walk‑through**
 -----------------------------------------------------------------------------
 This file implements three things:

   • an *Eisenstein integer* class  z = a0 + a1·ω  with  ω = e^{2πi/3}
   • an *exact* Euclidean division  quo_rem()  in the ring ℤ[ω]
   • the *Half‑GCD* algorithm, one of the fastest ways to compute a greatest
     common divisor when the inputs are very large
 """

from __future__ import annotations  # ► allow the class to refer to itself in type hints

import math
from typing import Tuple

# ---------------------------------------------------------------------------
# 1.  Tiny helper for the Euclidean division in ℤ[ω]
# ---------------------------------------------------------------------------

#  The six **unit directions** of the hexagonal lattice.  Graphically:
#     ↑ (0,1)
#  ↖       ↗
# (-1,1)   (1,0)
#  ↙       ↘
# (-1,0)   (1,-1)
#     ↓ (0,-1)
_NEIGHBOURS = ((1, 0), (0, 1), (-1, 1), (-1, 0), (0, -1), (1, -1))


def _round_nearest(z: int, d: int) -> int:
    """Symmetric integer rounding  ——  ⌊ (z + d/2) / d ⌋ for *any* sign of *z*.

    • *d* is strictly positive (we only call this with norms, which are > 0).
    • For   z ≥ 0   the formula above already works.
    • For   z <  0   we cannot add  +d/2  (would move us the wrong way),
      so we flip the sign, apply the positive rule, then flip back.
    """
    half = d >> 1  # bit‑shift right by 1  ⇒  divide by 2 (fast, exact)
    return (z + half) // d if z >= 0 else -((-z + half) // d)


# ---------------------------------------------------------------------------
# 2.  The EisensteinInteger class
# ---------------------------------------------------------------------------


class EisensteinInteger:
    """Eisenstein integers:  *hexagonal* analogue of Gaussian integers.

    Definition
    ----------
    ω  (pronounced "omega") is the complex cube‑root of 1 with positive
    imaginary part:  ω = e^{2πi/3} = −½ + i·√3/2.
    It satisfies  ω² + ω + 1 = 0  and  |ω| = 1.

    Every element of  ℤ[ω]  can be written uniquely as
        z = a₀ + a₁·ω,    with  a₀, a₁ ∈ ℤ.
    The pair (a₀, a₁) are the *hex‑coordinates* in the triangular lattice.

    In code we store just those two integers.
    """

    # Class‑level constants (created *after* the class body below)
    ZERO: "EisensteinInteger"  # 0 + 0·ω
    ONE: "EisensteinInteger"  # 1 + 0·ω

    # ────────────────────────────────────────────────────────────────────
    #  Construction & basic utilities
    # ────────────────────────────────────────────────────────────────────
    def __init__(self, a0: int = 0, a1: int = 0):
        if not isinstance(a0, int) or not isinstance(a1, int):
            raise TypeError("Coefficients a0 and a1 must be integers")
        self.a0 = a0
        self.a1 = a1

    def copy(self) -> "EisensteinInteger":
        """Clone *self* (handy for algorithms that mutate a temporary copy)."""
        return EisensteinInteger(self.a0, self.a1)

    def is_zero(self) -> bool:
        return self.a0 == 0 and self.a1 == 0

    def __str__(self) -> str:
        """Pretty print   3 + 2·ω,   −ω,  0,  …"""
        if self.a1 == 0:
            return str(self.a0)

        a0_part = f"{self.a0}" if self.a0 != 0 else ""
        # coefficient of ω
        if self.a1 == 1:
            a1_part = "ω"
        elif self.a1 == -1:
            a1_part = "-ω"
        else:
            a1_part = f"{self.a1}*ω"

        if a0_part == "":
            return a1_part
        sign = " + " if self.a1 > 0 else " - "
        abs_a1_part = "ω" if abs(self.a1) == 1 else f"{abs(self.a1)}*ω"
        return f"{a0_part}{sign}{abs_a1_part}"

    def __repr__(self) -> str:
        """Unambiguous form so that  eval(repr(z)) == z  (useful in a REPL)."""
        return f"EisensteinInteger({self.a0}, {self.a1})"

    # ────────────────────────────────────────────────────────────────────
    #  Equality & negation
    # ────────────────────────────────────────────────────────────────────
    def __eq__(self, other: object) -> bool:  # called by  z1 == z2
        return (
            isinstance(other, EisensteinInteger)
            and self.a0 == other.a0
            and self.a1 == other.a1
        )

    def __neg__(self) -> "EisensteinInteger":  # −z
        return EisensteinInteger(-self.a0, -self.a1)

    # ────────────────────────────────────────────────────────────────────
    #  Conjugation
    # ────────────────────────────────────────────────────────────────────
    def conjugate(self) -> "EisensteinInteger":
        """Field automorphism  ω ↦ ω²  (complex conjugation).
        Algebraically:  conj(a₀ + a₁·ω) = (a₀ − a₁) − a₁·ω.
        """
        return EisensteinInteger(self.a0 - self.a1, -self.a1)

    # ────────────────────────────────────────────────────────────────────
    #  Addition, subtraction, multiplication
    # ────────────────────────────────────────────────────────────────────
    def __add__(self, other: "EisensteinInteger") -> "EisensteinInteger":
        return EisensteinInteger(self.a0 + other.a0, self.a1 + other.a1)

    def __sub__(self, other: "EisensteinInteger") -> "EisensteinInteger":
        return EisensteinInteger(self.a0 - other.a0, self.a1 - other.a1)

    def __mul__(self, other: object) -> "EisensteinInteger":
        """Product formula derived from  (x₀ + x₁ω)(y₀ + y₁ω).
        Expand and use  ω² = −1 − ω.
        """
        if isinstance(other, EisensteinInteger):
            x0, x1, y0, y1 = self.a0, self.a1, other.a0, other.a1
            return EisensteinInteger(
                x0 * y0 - x1 * y1,  # coefficient of 1
                x0 * y1 + x1 * y0 - x1 * y1,  # coefficient of ω
            )
        elif isinstance(other, int):
            return EisensteinInteger(self.a0 * other, self.a1 * other)
        return NotImplemented

    # Support  int * EisensteinInteger  (commutative)
    __rmul__ = __mul__

    # ────────────────────────────────────────────────────────────────────
    #  Norm (always a *plain* integer)
    # ────────────────────────────────────────────────────────────────────
    def norm(self) -> int:
        """N(a₀ + a₁·ω)  =  a₀² + a₁² − a₀·a₁   ≥ 0
        Proof: treat ω as  −½ + i·√3/2  and compute |z|².
        """
        return self.a0 * self.a0 + self.a1 * self.a1 - self.a0 * self.a1

    # -------------------------------------------------------------------
    #  Euclidean division  ——  the heart of arithmetic in ℤ[ω]
    # -------------------------------------------------------------------
    def quo_rem(
        self, y: "EisensteinInteger"
    ) -> Tuple["EisensteinInteger", "EisensteinInteger"]:
        """Return  (q, r)  such that   self = q·y + r   with  N(r) < N(y).

        Strategy (all integer‑only):
          1. Multiply by the *conjugate* of y to get an element of the
             lattice that is colinear with y – this is the usual trick to
             mimic complex division without leaving the ring.
          2. Divide the two coordinates by the *positive* norm N(y) and
             apply symmetric rounding to obtain a *first guess* q.
          3. If this guess does not yet satisfy the Euclidean inequality
             N(r) < N(y), walk at most two unit steps in the hex lattice
             until it does.

        Because ℤ[ω] is a Euclidean domain, this procedure is guaranteed
        to terminate and the while‑loop never needs more than two laps.
        """
        if y.is_zero():
            raise ZeroDivisionError("division by zero EisensteinInteger")

        nrm = y.norm()  # denominator of the coordinates ( > 0 )
        num = self * y.conjugate()  # numerator in the same lattice line

        # --- 1st approximation by independent rounding -----------------
        q0 = _round_nearest(num.a0, nrm)
        q1 = _round_nearest(num.a1, nrm)
        q = EisensteinInteger(q0, q1)
        r = self - q * y

        # --- Improve the quotient until Euclidean property holds --------
        while r.norm() >= nrm:  # very rarely true (≈ 1 in 12 cases)
            best_q, best_r, best_norm = None, None, r.norm()
            # test the 6 neighbours
            for dp, dq in _NEIGHBOURS:
                cand_q = EisensteinInteger(q0 + dp, q1 + dq)
                cand_r = self - cand_q * y
                cand_norm = cand_r.norm()
                if cand_norm < best_norm:
                    best_q, best_r, best_norm = cand_q, cand_r, cand_norm
            # move to the better neighbour (Euclidean property assures one exists)
            q, r = best_q, best_r
            q0, q1 = q.a0, q.a1  # update centre for the *next* neighbour pass
        return q, r

    # Allow the syntactic sugar  z // y   and  z % y -------------------------
    def __floordiv__(self, other: "EisensteinInteger") -> "EisensteinInteger":
        return self.quo_rem(other)[0]

    def __mod__(self, other: "EisensteinInteger") -> "EisensteinInteger":
        return self.quo_rem(other)[1]


# -------- class constants (must be assigned *after* the class is complete) ---
EisensteinInteger.ZERO = EisensteinInteger(0, 0)
EisensteinInteger.ONE = EisensteinInteger(1, 0)

# ---------------------------------------------------------------------------
# 3.  Half‑GCD  ——  sub‑quadratic gcd via size halving
# ---------------------------------------------------------------------------


def half_gcd(
    a: EisensteinInteger,
    b: EisensteinInteger,
) -> Tuple[EisensteinInteger, EisensteinInteger, EisensteinInteger]:
    """Return  (w, v, u)  with  w = a·u + b·v  and  N(w) < √N(a).

    Why is that useful?  The classical Euclidean gcd repeatedly replaces the
    *larger* input by its remainder modulo the smaller one – this shrinks one
    argument at every step, but only by a *constant* factor, so it needs
    O(log N) iterations.

    **Half‑GCD** knocks out *half* the size (in bits) at every major step and
    therefore runs in roughly O(log N / log log N) time.

    The idea is to run the Euclidean algorithm *just long enough* for the
    remainder’s norm to dip below √N(a).  The output coefficients u, v let us
    continue assembling the gcd later (the full gcd algorithm calls half_gcd
    recursively on smaller and smaller inputs).
    """
    if not isinstance(a, EisensteinInteger) or not isinstance(b, EisensteinInteger):
        raise TypeError("Inputs must be EisensteinInteger instances")

    # Extended Euclidean bookkeeping matrices
    a_run, b_run = a.copy(), b.copy()
    u, v = EisensteinInteger.ONE.copy(), EisensteinInteger.ZERO.copy()
    u_, v_ = EisensteinInteger.ZERO.copy(), EisensteinInteger.ONE.copy()
    # Invariants throughout the loop:
    #     a_run =  a·u  + b·v
    #     b_run =  a·u_ + b·v_

    limit = math.isqrt(a.norm())  # √N(a)  (integer square root)

    # Safety valve: more than ~2e5 iterations would imply a bug
    for _ in range(200_000):
        if b_run.norm() < limit or b_run.is_zero():
            break
        q, r = a_run.quo_rem(b_run)

        # update (a_run, b_run)  and the coefficient matrix simultaneously
        a_run, b_run = b_run, r
        u, u_ = u_, u - q * u_
        v, v_ = v_, v - q * v_
    else:
        raise RuntimeError("Half‑GCD failed to converge – too many steps")

    # *Return order* follows the classic Half‑GCD papers:  (w, v, u)
    return b_run, v_, u_


if __name__ == "__main__":
    z1 = EisensteinInteger(3, 2)  # 3 + 2ω
    z2 = EisensteinInteger(1, -1)  # 1 - ω

    print(f"z1 = {z1}")
    print(f"z2 = {z2}")
    print(f"z1 + z2 = {z1 + z2}")
    print(f"z1 - z2 = {z1 - z2}")
    print(f"z1 * z2 = {z1 * z2}")
    print(f"Norm(z1) = {z1.norm()}")
    print(f"Conj(z1) = {z1.conjugate()}")
    print(f"-z1 = {-z1}")

    q, r = z1.quo_rem(z2)
    print(f"z1 // z2 = {q}")
    print(f"z1 % z2 = {r}")
    print(f"Check: z2 * q + r = {z2 * q + r}, Expected: {z1}")

    # Half GCD example
    a = EisensteinInteger(10, 3)
    b = EisensteinInteger(3, 1)
    print(f"\nHalf GCD for a = {a}, b = {b}")
    w, v_res, u_res = half_gcd(a, b)
    print(f"  w = {w}")
    print(f"  v = {v_res}")
    print(f"  u = {u_res}")
    check = a * u_res + b * v_res
    print(f"  Check: a*u + b*v = {check}, Expected w: {w}")
    print(f"  Norm(w) = {w.norm()}, Limit = {math.isqrt(a.norm())}")

    # a = EisensteinInteger(8+7j * (math.sqrt(3)/2 + 0.5j)) # Approximate large number
    a_int = EisensteinInteger(13, 14)  # Example large int coeffs
    b_int = EisensteinInteger(5, 3)
    print(f"\nHalf GCD for a = {a_int}, b = {b_int}")
    w, v_res, u_res = half_gcd(a_int, b_int)
    print(f"  w = {w}")
    print(f"  v = {v_res}")
    print(f"  u = {u_res}")
    check = a_int * u_res + b_int * v_res
    print(f"  Check: a*u + b*v = {check}, Expected w: {w}")
    print(f"  Norm(w) = {w.norm()}, Limit = {math.isqrt(a_int.norm())}")

File: tools/garaga_rs/src/hints/eisenstein.rs
use num_bigint::BigInt;
use num_traits::{One, Signed, Zero};

/// Axial coordinates of the 6 unit directions in the hexagonal lattice.
const NEIGHBOURS: [(i32, i32); 6] = [(1, 0), (0, 1), (-1, 1), (-1, 0), (0, -1), (1, -1)];

/// Symmetric integer rounding  ⌊(z + d/2) / d⌋  valid for *any* sign of `z`.
fn round_nearest(z: &BigInt, d: &BigInt) -> BigInt {
    debug_assert!(*d > BigInt::zero());
    let half = d >> 1; // d / 2
    if *z >= BigInt::zero() {
        (z + &half) / d
    } else {
        let quotient: BigInt = ((-z) + &half) / d;
        -quotient
    }
}

/*
Represents an Eisenstein integer z = a0 + a1*ω, where ω = exp(2πi/3).

ω satisfies ω^2 + ω + 1 = 0.
Uses Python's built-in arbitrary-precision integers.

Attributes:
    a0: The integer coefficient of 1.
    a1: The integer coefficient of ω.
*/
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct EisensteinInteger {
    pub a0: BigInt,
    pub a1: BigInt,
}

impl EisensteinInteger {
    /*Initializes an EisensteinInteger.*/
    pub fn new(a0: BigInt, a1: BigInt) -> Self {
        Self { a0, a1 }
    }
    /*Sets the value of self to the value of other. Returns self.*/
    pub fn set(&mut self, other: &Self) -> &Self {
        self.a0 = other.a0.clone();
        self.a1 = other.a1.clone();
        self
    }
    /*Sets the value of self to zero. Returns self.*/
    pub fn set_zero(&mut self) -> &Self {
        self.a0 = BigInt::zero();
        self.a1 = BigInt::zero();
        self
    }
    /*Sets the value of self to one. Returns self.*/
    pub fn set_one(&mut self) -> &Self {
        self.a0 = BigInt::one();
        self.a1 = BigInt::zero();
        self
    }
    /*
    Computes the complex conjugate.

    conj(a0 + a1*ω) = (a0 - a1) - a1*ω
    */
    pub fn conjugate(&self) -> Self {
        Self::new(&self.a0 - &self.a1, -&self.a1)
    }
    /*
    Computes the norm N(self).

    N(a0 + a1*ω) = a0^2 + a1^2 - a0*a1
    The norm is always non-negative.
    */
    pub fn norm(&self) -> BigInt {
        &self.a0.pow(2) + &self.a1.pow(2) - &self.a0 * &self.a1
    }
    /*
    Performs Euclidean division self / y using rounding to the NEAREST Eisenstein integer.

    This method guarantees norm(remainder) < norm(divisor).

    Args:
        y: The divisor (must be non-zero).

    Returns:
        A tuple (quotient, remainder).

    Raises:
        ZeroDivisionError: If y is zero.
        TypeError: If y is not an EisensteinInteger.
    */
    /* ---------- Euclidean division in ℤ[ω] ---------- */
    pub fn quo_rem(&self, y: &Self) -> Result<[Self; 2], String> {
        if y.is_zero() {
            return Err("division by zero EisensteinInteger".into());
        }
        let nrm = y.norm(); // positive
        let num = self * &y.conjugate(); // still integer-only

        // first guess by independent symmetric rounding
        let mut q0 = round_nearest(&num.a0, &nrm);
        let mut q1 = round_nearest(&num.a1, &nrm);
        let mut q = Self::new(q0.clone(), q1.clone());
        // Calculate product first, then subtract its reference
        let qy = &q * y;
        let mut r = self - &qy;

        // walk the neighbourhood until  N(r) < N(y)  (≤ 2 iterations)
        while r.norm() >= nrm {
            let mut best_q = None;
            let mut best_r = None;
            let mut best_n2 = r.norm(); // current (worst) remainder

            for (dp, dq) in NEIGHBOURS.iter() {
                let cand_q = Self::new(&q0 + dp, &q1 + dq);
                // Calculate product first, then subtract its reference
                let cand_qy = &cand_q * y;
                let cand_r = self - &cand_qy;
                let cand_n2 = cand_r.norm();
                if cand_n2 < best_n2 {
                    best_q = Some(cand_q);
                    best_r = Some(cand_r);
                    best_n2 = cand_n2;
                }
            }
            q = best_q.expect("Euclidean property violated");
            r = best_r.unwrap();
            q0 = q.a0.clone();
            q1 = q.a1.clone(); // centre for 2nd lap
        }
        Ok([q, r])
    }
}

impl std::fmt::Display for EisensteinInteger {
    /*Returns a user-friendly string representation (e.g., '3 + 2*ω').*/
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        if self.a1.is_zero() {
            write!(f, "{}", self.a0)
        } else {
            let a0_str = if !self.a0.is_zero() {
                format!("{}", self.a0)
            } else {
                "".into()
            };

            let a1_str = if self.a1 == BigInt::one() {
                "ω".into()
            } else if self.a1 == (-1).into() {
                "-ω".into()
            } else {
                format!("{}*ω", self.a1)
            };

            if a0_str.is_empty() {
                write!(f, "{}", a1_str)
            } else {
                let sign = if self.a1 > BigInt::zero() {
                    " + "
                } else {
                    " - "
                };
                let a1_abs_str = if self.a1.abs() == BigInt::one() {
                    "ω".into()
                } else {
                    format!("{}*ω", self.a1.abs())
                };
                write!(f, "{}{}{}", a0_str, sign, a1_abs_str)
            }
        }
    }
}

impl std::ops::Neg for EisensteinInteger {
    type Output = Self;

    /*Computes the negation -self.*/
    fn neg(self) -> Self::Output {
        Self {
            a0: -self.a0,
            a1: -self.a1,
        }
    }
}

impl std::ops::Add for EisensteinInteger {
    type Output = Self;

    /*Computes the sum self + other.*/
    fn add(self, other: Self) -> Self::Output {
        &self + &other
    }
}

impl<'a> std::ops::Add<&'a EisensteinInteger> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*Computes the sum self + other.*/
    fn add(self, other: Self) -> Self::Output {
        EisensteinInteger::new(&self.a0 + &other.a0, &self.a1 + &other.a1)
    }
}

impl std::ops::Sub for EisensteinInteger {
    type Output = Self;

    /*Computes the difference self - other.*/
    fn sub(self, other: Self) -> Self::Output {
        &self - &other
    }
}

impl<'a> std::ops::Sub<&'a EisensteinInteger> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*Computes the difference self - other.*/
    fn sub(self, other: Self) -> Self::Output {
        EisensteinInteger::new(&self.a0 - &other.a0, &self.a1 - &other.a1)
    }
}

impl std::ops::Mul for EisensteinInteger {
    type Output = Self;

    /*
    Computes the product self * other.

    Supports multiplication by another EisensteinInteger.
    Uses the identity (x0+x1ω)(y0+y1ω) = (x0y0 - x1y1) + (x0y1 + x1y0 - x1y1)ω.
    */
    fn mul(self, other: Self) -> Self::Output {
        &self * &other
    }
}

impl<'a> std::ops::Mul<&'a EisensteinInteger> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*
    Computes the product self * other.

    Supports multiplication by another EisensteinInteger.
    Uses the identity (x0+x1ω)(y0+y1ω) = (x0y0 - x1y1) + (x0y1 + x1y0 - x1y1)ω.
    */
    fn mul(self, other: Self) -> Self::Output {
        let (x0, x1) = (&self.a0, &self.a1);
        let (y0, y1) = (&other.a0, &other.a1);
        let res_a0 = x0 * y0 - x1 * y1;
        let res_a1 = x0 * y1 + x1 * y0 - x1 * y1;
        EisensteinInteger::new(res_a0, res_a1)
    }
}

impl std::ops::Mul<BigInt> for EisensteinInteger {
    type Output = Self;

    /*Computes the product self * int.*/
    fn mul(self, other: BigInt) -> Self::Output {
        &self * &other
    }
}

impl<'a> std::ops::Mul<&'a BigInt> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*Computes the product self * int.*/
    fn mul(self, other: &BigInt) -> Self::Output {
        other * self
    }
}

impl std::ops::Mul<EisensteinInteger> for BigInt {
    type Output = EisensteinInteger;

    /*Computes the product int * self.*/
    fn mul(self, other: EisensteinInteger) -> Self::Output {
        &self * &other
    }
}

impl<'a> std::ops::Mul<&'a EisensteinInteger> for &'a BigInt {
    type Output = EisensteinInteger;

    /*Computes the product int * self.*/
    fn mul(self, other: &EisensteinInteger) -> Self::Output {
        EisensteinInteger::new(self * &other.a0, self * &other.a1)
    }
}

impl std::ops::Div for EisensteinInteger {
    type Output = Self;

    /*Computes the quotient self // other using Euclidean rounding division.*/
    fn div(self, other: Self) -> Self::Output {
        &self / &other
    }
}

impl<'a> std::ops::Div<&'a EisensteinInteger> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*Computes the quotient self // other using Euclidean rounding division.*/
    fn div(self, other: Self) -> Self::Output {
        let [q, _] = self.quo_rem(other).unwrap();
        q
    }
}

impl std::ops::Rem for EisensteinInteger {
    type Output = Self;

    /*Computes the remainder self % other using Euclidean rounding division.*/
    fn rem(self, other: Self) -> Self::Output {
        &self % &other
    }
}

impl<'a> std::ops::Rem<&'a EisensteinInteger> for &'a EisensteinInteger {
    type Output = EisensteinInteger;

    /*Computes the remainder self % other using Euclidean rounding division.*/
    fn rem(self, other: Self) -> Self::Output {
        let [_, r] = self.quo_rem(other).unwrap();
        r
    }
}

impl Zero for EisensteinInteger {
    fn zero() -> Self {
        Self::new(BigInt::zero(), BigInt::zero())
    }

    /*Checks if the value is zero.*/
    fn is_zero(&self) -> bool {
        self.a0.is_zero() && self.a1.is_zero()
    }
}

impl One for EisensteinInteger {
    fn one() -> Self {
        Self::new(BigInt::one(), BigInt::zero())
    }
}

/*
Computes the partial GCD result using the Half-GCD approach based on norm reduction.

Args:
    a: The first EisensteinInteger.
    b: The second EisensteinInteger.

Returns:
    A tuple (w, v, u) such that w = a*u + b*v and norm(w) is approximately
    less than sqrt(norm(a)). The identity w = a*u + b*v always holds.

Raises:
    TypeError: If inputs are not EisensteinInteger instances.
    ValueError: If the norm of 'a' is unexpectedly negative.
    RuntimeError: If the algorithm takes an excessive number of iterations.
*/
pub fn half_gcd(
    a: &EisensteinInteger,
    b: &EisensteinInteger,
) -> Result<[EisensteinInteger; 3], String> {
    // Initialize variables for the Extended Euclidean Algorithm
    let mut a_run = a.clone();
    let mut b_run = b.clone();
    let mut u = EisensteinInteger::one(); // Use class constant copies
    let mut v = EisensteinInteger::zero();
    let mut u_ = EisensteinInteger::zero();
    let mut v_ = EisensteinInteger::one();
    // Invariants: a_run = a*u + b*v,  b_run = a*u_ + b*v_

    let norm_a = a.norm();
    if norm_a < BigInt::zero() {
        return Err("Norm of input 'a' cannot be negative".into());
    }
    // Calculate the termination threshold
    let limit = norm_a.sqrt();

    // Loop while the norm of the 'smaller' value (b_run) is >= the limit
    let mut iteration = 0;
    let max_iterations = 20000; // Safety limit

    while b_run.norm() >= limit {
        if iteration > max_iterations {
            return Err(format!(
                "HalfGCD exceeded {} iterations for a={}, b={}",
                max_iterations, a, b
            ));
        }

        if b_run.is_zero() {
            break; // GCD is a_run
        }

        let [quotient, remainder] = match a_run.quo_rem(&b_run) {
            Ok(result) => result,
            _ => break, // Should ideally not happen if b_run.is_zero() check works
        };

        // Update coefficients
        let next_u_ = u - &quotient * &u_;
        let next_v_ = v - &quotient * &v_;

        // Update state
        (a_run, b_run) = (b_run, remainder);
        (u, u_) = (u_, next_u_);
        (v, v_) = (v_, next_v_);

        iteration += 1;
    }

    // Return the final state corresponding to the Go implementation's return values
    // (b_run, v_, u_) which satisfy b_run = a*u_ + b*v_
    Ok([b_run, v_, u_])
}

#[allow(unused_imports)]
mod tests {
    use super::*;

    #[test]
    fn test_01() -> Result<(), String> {
        let z1 = EisensteinInteger::new(3.into(), 2.into()); // 3 + 2ω
        let z2 = EisensteinInteger::new(1.into(), (-1).into()); // 1 - ω

        println!("z1 = {}", z1);
        println!("z2 = {}", z2);
        println!("z1 + z2 = {}", &z1 + &z2);
        println!("z1 - z2 = {}", &z1 - &z2);
        println!("z1 * z2 = {}", &z1 * &z2);
        println!("Norm(z1) = {}", z1.norm());
        println!("Conj(z1) = {}", z1.conjugate());
        println!("-z1 = {}", -z1.clone());

        let [q, r] = z1.quo_rem(&z2)?;
        println!("z1 // z2 = {}", q);
        println!("z1 % z2 = {}", r);
        println!("Check: z2 * q + r = {}, Expected: {}", &(&z2 * &q) + &r, z1);
        assert_eq!(&(&z2 * &q) + &r, z1);

        // Half GCD example
        let a = EisensteinInteger::new(10.into(), 3.into());
        let b = EisensteinInteger::new(3.into(), 1.into());
        println!("\nHalf GCD for a = {}, b = {}", a, b);
        let [w, v_res, u_res] = half_gcd(&a, &b)?;
        println!("  w = {}", w);
        println!("  v = {}", v_res);
        println!("  u = {}", u_res);
        let check = &a * &u_res + &b * &v_res;
        println!("  Check: a*u + b*v = {}, Expected w: {}", check, w);
        println!("  Norm(w) = {}, Limit = {}", w.norm(), a.norm().sqrt());
        assert_eq!(check, w);

        // a = EisensteinInteger(8+7j * (math.sqrt(3)/2 + 0.5j)) # Approximate large number
        let a_int = EisensteinInteger::new(13.into(), 14.into()); // Example large int coeffs
        let b_int = EisensteinInteger::new(5.into(), 3.into());
        println!("\nHalf GCD for a = {}, b = {}", a_int, b_int);
        let [w, v_res, u_res] = half_gcd(&a_int, &b_int)?;
        println!("  w = {}", w);
        println!("  v = {}", v_res);
        println!("  u = {}", u_res);
        let check = &a_int * &u_res + &b_int * &v_res;
        println!("  Check: a*u + b*v = {}, Expected w: {}", check, w);
        println!("  Norm(w) = {}, Limit = {}", w.norm(), a_int.norm().sqrt());
        assert_eq!(check, w);

        Ok(())
    }
}
